test/core/event_engine/thread_pool_test.cc

```diff
@@ -12,117 +12,204 @@
-
-#include <stdlib.h>
+#include "src/core/lib/event_engine/thread_pool/thread_pool.h"
 
+#include <atomic>
 #include <chrono>
+#include <cmath>
+#include <functional>
 #include <thread>
 
+#include "absl/time/clock.h"
+#include "absl/time/time.h"
 #include "gtest/gtest.h"
 
-#include <grpc/support/log.h>
+#include <grpc/grpc.h>
 
+#include "src/core/lib/event_engine/thread_pool/original_thread_pool.h"
+#include "src/core/lib/event_engine/thread_pool/work_stealing_thread_pool.h"
 #include "src/core/lib/gprpp/notification.h"
+#include "test/core/util/test_config.h"
 
 namespace grpc_event_engine {
 namespace experimental {
 
-TEST(ThreadPoolTest, CanRunClosure) {
-  ThreadPool p;
+template <typename T>
+class ThreadPoolTest : public testing::Test {};
+
+using ThreadPoolTypes =
+    ::testing::Types<OriginalThreadPool, WorkStealingThreadPool>;
+TYPED_TEST_SUITE(ThreadPoolTest, ThreadPoolTypes);
+
+TYPED_TEST(ThreadPoolTest, CanRunAnyInvocable) {
+  TypeParam p(8);
   grpc_core::Notification n;
   p.Run([&n] { n.Notify(); });
   n.WaitForNotification();
   p.Quiesce();
 }
 
-TEST(ThreadPoolTest, CanDestroyInsideClosure) {
-  auto p = std::make_shared<ThreadPool>();
+TYPED_TEST(ThreadPoolTest, CanDestroyInsideClosure) {
+  auto* p = new TypeParam(8);
   grpc_core::Notification n;
   p->Run([p, &n]() mutable {
-    std::this_thread::sleep_for(std::chrono::seconds(1));
     // This should delete the thread pool and not deadlock
     p->Quiesce();
-    p.reset();
+    delete p;
     n.Notify();
   });
-  // Make sure we're not keeping the thread pool alive from outside the loop
-  p.reset();
   n.WaitForNotification();
 }
 
-TEST(ThreadPoolTest, CanSurviveFork) {
-  ThreadPool p;
-  grpc_core::Notification n;
-  gpr_log(GPR_INFO, "run callback 1");
-  p.Run([&n, &p] {
+TYPED_TEST(ThreadPoolTest, CanSurviveFork) {
+  TypeParam p(8);
+  grpc_core::Notification inner_closure_ran;
+  p.Run([&inner_closure_ran, &p] {
     std::this_thread::sleep_for(std::chrono::seconds(1));
-    gpr_log(GPR_INFO, "run callback 2");
-    p.Run([&n] {
+    p.Run([&inner_closure_ran] {
       std::this_thread::sleep_for(std::chrono::seconds(1));
-      gpr_log(GPR_INFO, "notify");
-      n.Notify();
+      inner_closure_ran.Notify();
     });
   });
-  gpr_log(GPR_INFO, "prepare fork");
+  // simulate a fork and watch the child process
   p.PrepareFork();
-  gpr_log(GPR_INFO, "postfork child");
   p.PostforkChild();
-  n.WaitForNotification();
+  inner_closure_ran.WaitForNotification();
   grpc_core::Notification n2;
-  gpr_log(GPR_INFO, "run callback 3");
-  p.Run([&n2] {
-    gpr_log(GPR_INFO, "notify");
-    n2.Notify();
-  });
-  gpr_log(GPR_INFO, "wait for notification");
+  p.Run([&n2] { n2.Notify(); });
   n2.WaitForNotification();
   p.Quiesce();
 }
 
-void ScheduleSelf(ThreadPool* p) {
-  p->Run([p] { ScheduleSelf(p); });
+TYPED_TEST(ThreadPoolTest, ForkStressTest) {
+  // Runs a large number of closures and multiple simulated fork events,
```

Comment:
Can you add more comments for this test ? Is it testing that a Fork operation is not blocked indefinitely ?

Reply:
Done